samples/packages/awscdk-typescript/src/package-config.ts

```diff
@@ -83,18 +82,27 @@ export enum ComponentBlockPublicAccess {
   DEFAULT = BLOCK_ALL,
 }
 
+export enum ComponentTyping {
+  MESSAGE_TOPIC = 'MESSAGE_TOPIC',
+  MESSAGE_QUEUE = 'MESSAGE_QUEUE',
+  STORE_OBJECT = 'STORE_OBJECT',
+  DISTRIBUTION_WEBSITE = 'DISTRIBUTION_WEBSITE',
+  KEY_SOFTWARE = 'KEY_SOFTWARE',
+}
+
 export class PackageComponentKeySoftwareConfig implements PackageComponentConfig {
+  static TYPING = (manifest as PackageManifest).constructs.components.types[ComponentTyping.KEY_SOFTWARE];
 
   constructor(
     public readonly name: string,
     public readonly spec?: ComponentKeySpec,
     public readonly usage?: ComponentKeyUsage,
     public readonly enabled?: boolean,
     public readonly rotation?: boolean,
-  ) { }
+  ) {}
 
   public static has(config: cdf.TraitConfig<PackageComponentConfig>): boolean {
-    return config.type == 'key' && config.subtype == 'software';
+    return config.type == this.TYPING.type && config.subtype == this.TYPING.subtype;
```

Comment:
A new pattern: 
1. defining types of components by their keys only
2. looking up from `cdk.manifest.json` using a key to get details such as type and subtype values
3. comparing with what's defined in config json file to decide whether a component is matching